File: VulkanEngineDLL/Material.cpp
#include "Material.h"
#include "json.h"

Material Material_CreateMaterial(const RendererStateDLL& rendererStateDLL, int bufferIndex, VulkanBuffer& materialBuffer, const char* jsonString)
{
    const RendererState renderer = VulkanRenderer_ConvertToVulkanRenderer(rendererStateDLL);
    materialBuffer = VulkanBuffer_CreateVulkanBuffer(renderer, bufferIndex, sizeof(MaterialProperitiesBuffer), 1, BufferTypeEnum::BufferType_MaterialProperitiesBuffer, VK_BUFFER_USAGE_STORAGE_BUFFER_BIT,
                                                                                                                                                                       VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
                                                                                                                                                                       VK_MEMORY_PROPERTY_HOST_COHERENT_BIT |
                                                                                                                                                                       VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT, false);
    nlohmann::json json = Json::ReadJson(jsonString);
    return Material
    {
        .materialGuid = VkGuid(json["MaterialId"].get<String>().c_str()),
        .ShaderMaterialBufferIndex = 0,
        .MaterialBufferId = bufferIndex,
        .AlbedoMapId = json["AlbedoMapId"].get<std::string>().empty() ? VkGuid() : json["AlbedoMapId"].get<std::string>().c_str(),
        .MetallicRoughnessMapId = json["MetallicRoughnessMapId"].get<std::string>().empty() ? VkGuid() : json["MetallicRoughnessMapId"].get<std::string>().c_str(),
        .MetallicMapId = json["MetallicMapId"].get<std::string>().empty() ? VkGuid() : json["MetallicMapId"].get<std::string>().c_str(),
        .RoughnessMapId = json["RoughnessMapId"].get<std::string>().empty() ? VkGuid() : json["RoughnessMapId"].get<std::string>().c_str(),
        .AmbientOcclusionMapId = json["AmbientOcclusionMapId"].get<std::string>().empty() ? VkGuid() : json["AmbientOcclusionMapId"].get<std::string>().c_str(),
        .NormalMapId = json["NormalMapId"].get<std::string>().empty() ? VkGuid() : json["NormalMapId"].get<std::string>().c_str(),
        .DepthMapId = json["DepthMapId"].get<std::string>().empty() ? VkGuid() : json["DepthMapId"].get<std::string>().c_str(),
        .AlphaMapId = json["AlphaMapId"].get<std::string>().empty() ? VkGuid() : json["AlphaMapId"].get<std::string>().c_str(),
        .EmissionMapId = json["EmissionMapId"].get<std::string>().empty() ? VkGuid() : json["EmissionMapId"].get<std::string>().c_str(),
        .HeightMapId = json["HeightMapId"].get<std::string>().empty() ? VkGuid() : json["HeightMapId"].get<std::string>().c_str(),
        .Albedo = vec3(json["Albedo"][0], json["Albedo"][1], json["Albedo"][2]),
        .Emission = vec3(json["Emission"][0], json["Emission"][1], json["Emission"][2]),
        .Metallic = json["Metallic"],
        .Roughness = json["Roughness"],
        .AmbientOcclusion = json["AmbientOcclusion"],
        .Alpha = json["Alpha"],
    };
}

void Material_UpdateBuffer(RendererStateDLL& rendererStateDLL, VulkanBuffer& materialBuffer, MaterialProperitiesBuffer& materialProperties)
{
    const RendererState renderer = VulkanRenderer_ConvertToVulkanRenderer(rendererStateDLL);
    VulkanBuffer_UpdateBufferMemory(renderer, materialBuffer, static_cast<void*>(&materialProperties), sizeof(MaterialProperitiesBuffer), 1);
    VulkanRenderer_DeleteVulkanRenderStatePtrs(&rendererStateDLL);
}

void Material_DestroyBuffer(const RendererStateDLL& rendererStateDLL, VulkanBuffer& materialBuffer)
{
    const RendererState renderer = VulkanRenderer_ConvertToVulkanRenderer(rendererStateDLL);
    VulkanBuffer_DestroyBuffer(renderer, materialBuffer);
}

File: VulkanGameEngineLevelEditor/GameEngine/Structs/Material.cs
using GlmSharp;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace VulkanGameEngineLevelEditor.GameEngine.Structs
{
    public struct Material
    {
        public int VectorMapKey;
        public Guid MaterialId { get; set; } = new Guid();
        public uint ShaderMaterialBufferIndex { get; set; } = 0;
        public int MaterialBufferId { get; set; } = 0;

        public Guid AlbedoMapId { get; set; } = new Guid();
        public Guid MetallicRoughnessMapId { get; set; } = new Guid();
        public Guid MetallicMapId { get; set; } = new Guid();
        public Guid RoughnessMapId { get; set; } = new Guid();
        public Guid AmbientOcclusionMapId { get; set; } = new Guid();
        public Guid NormalMapId { get; set; } = new Guid();
        public Guid DepthMapId { get; set; } = new Guid();
        public Guid AlphaMapId { get; set; } = new Guid();
        public Guid EmissionMapId { get; set; } = new Guid();
        public Guid HeightMapId { get; set; } = new Guid();

        public vec3 Albedo { get; set; } = new vec3(0.0f, 0.35f, 0.45f);
        public vec3 Emission { get; set; } = new vec3(0.0f);
        public float Metallic { get; set; } = 0.0f;
        public float Roughness { get; set; } = 0.0f;
        public float AmbientOcclusion { get; set; } = 1.0f;
        public float Alpha { get; set; } = 1.0f;

        public Material()
        {
        }
    };
}
